### PR TITLE
Update NormalizeExpressionLevels_allsampleref- Add datatype specific normalization

### PR DESCRIPTION
**Pull request by rmadupuri**
**RNA-Seq data:** z-scores on original data or log-transformed data
   - Drop any non-numeric values like NA, NaN, Null..
   - And drop zero's and negative values.
  
**Microarray Expression or RPPA:** z-scores on original data or log-transformed data
   - Drop any non-numeric values like NA, NaN, Null..
   - Do not drop zero’s or negative values.

**Changes made to the script:**
- Added a new argument that takes in the input data_type (`microarray`, `rppa` or `rnaseq`) and performs the respective normalization.

Here's how we are handling the Zero and Negative values when log transforming:

Replace the negative values to 0 and add a pseudo-count (+1) to all observed counts prior to applying log transform.
example,
- if raw value is -1, the log transform would be log(0+1)
- if the value is 0, the log transform would be log(0+1)
- if the value is 1, the log transform would be log(1+1)

The log value will be zero when the original value is zero or negative.

**Affected Data:** As we initially had a single script for all expression data that dropped zero's,
- Microarray Expression profiles: genes with -ve or zero raw expression value (Redo the z-scores with zero/-ve values included)
- Maybe RPPA: genes with zero rppa raw value. No zero counts observed in most of the RPPA profiles.

### CODE REVIEW

**Comment by jjgao** [comment]
Thanks, @rmadupuri. This looks good based on the description. I didn't carefully review the code change though...

For the new argument (microarray, rppa or rnaseq), I am wondering if it's better to just have a binary argument to indicate whether 0s should be excluded and leave the logic of file type to the data pipeline?

**Comment by rmadupuri** [comment]
Thank you for the review @jjgao. I have updated the script to include an optional argument to indicate the exclusion of zero's instead of passing the file type.

One last question, for cases where we exclude zero's and negative values what would the zscore value for them be? Do we calculate the zscore using (r-mu)/sigma or should they be NA. Thank you!

File: zscores/zscores_relative_allsamples/NormalizeExpressionLevels_allsampleref.py [hunk 1]
```diff
@@ -52,26 +52,34 @@ def calculate_mean(data,n):
...
+			try:
+				val = float(item)
+				if val > 0:
+					filtered_expression_values.append(val)
+			except:
+				continue
+	else:
+		for item in expression_values:
+			try:
+				val = float(item)
```
Comment (sheridancbio) [inline zscores/zscores_relative_allsamples/NormalizeExpressionLevels_allsampleref.py:79]: now zeros (and negative numbers) are used here

File: zscores/zscores_relative_allsamples/NormalizeExpressionLevels_allsampleref.py [hunk 2]
```diff
@@ -128,8 +134,11 @@ def find_sample_position(infile):
 		sample_position = 1
 	if header[1] in HEADER_KEYWORDS:
 		sample_position = 2
+	if header[0] not in HEADER_KEYWORDS and header[1] in HEADER_KEYWORDS:
+		print("ERROR: The first column must be a gene identifier. It should be either one of Hugo_Symbol, Entrez_Gene_Id or Composite.Element.REF\nExiting..")
```
Comment (sheridancbio) [inline zscores/zscores_relative_allsamples/NormalizeExpressionLevels_allsampleref.py:138]: extra check was added here

File: zscores/zscores_relative_allsamples/NormalizeExpressionLevels_allsampleref.py [hunk 3]
```diff
@@ -152,76 +161,80 @@ def log_transform_data(input_filename,sample_position):
 			elif line.startswith('Composite.Element.REF') or line.startswith('Hugo_Symbol') or line.startswith('Entrez_Gene_Id'):
 				log_transformed_data += line
 			else:
-				cols = line.rstrip('\n').rstrip('\r').split('\t')
-				scores = cols[sample_position:]
-				transformed_cols = []
-				transformed_cols.extend(cols[:sample_position])
+				data = line.rstrip('\n').rstrip('\r').split('\t')
+				raw_values = data[sample_position:]
+				transformed_values = data[:sample_position]
```
Comment (sheridancbio) [inline zscores/zscores_relative_allsamples/NormalizeExpressionLevels_allsampleref.py:166]: now we keep them separate

**Review by sheridancbio — APPROVED** [review]
The code structure looks good, and documentation seems clear. The statistical correctness of these changes is not part of my review.

**Pull request merged**